# The Indiana Training Program in Public and Population Health Informatics


## Exercise 2

*Competencies addressed*   

1. Apply data merging/linking and reshaping methods  (I.1.3)
2. Develop an understanding of processing different kinds of data (e.g. string processing)  (I.1.5)
3. Demonstrate the ability to describe data (find means, standard deviations, outliers, evaluate correlations etc.) (II.1.2)

*Learning objectives*  
 At the end of this module the student will be able to do the following:

1. Set up the working environment in R
2. Read the dataset prior to loading it into R
3. Import a text delimited dataset into R
4. Run summary statistics
5. Make a dataset more meaningful by linking the multiple datasets
6. Import a stata file into R
7. Understand the significance of choosing the right variable type when importing a dataset into R


This exercises uses the medication data from the Type-2 Diabetes dataset. The medication data contains detailed information on the various medications prescribed to the patients. Additional information on dosage and when the medications were prescribed is also available. We first start by loading the data into R. Once the dataset is loaded we use `head` to view the first few rows to get an idea of what this dataset looks like.

In [1]:
getwd()  #check current directory
setwd("/N/dc2/projects/T15/Sample") # setting working folder
medication <- read.csv("medication.csv", header= TRUE)
head(medication)
summary(medication)

[1] "/N/dc2/projects/T15/Sample"

STUDYID DAYS_MED_INDEX DRUG_NAME              NDC_CODE    STRENGTH
1 103     2757           GLIPIZIDE 5 MG TABLET  51079081020 NA      
2 239     6871           NOVOLIN N                169183411 NA      
3 239      811           HUMULIN N                  2831501 NA      
4 239     3751           HUMULIN N    INJ U-100     2831501 NA      
5 239     3819           HUMULIN R    INJ U-100     2821501 NA      
6 239     1128           HUMULIN N                  2831501 NA      
  DISPENSE_AMOUNT NUMBER_OF_DAYS_SUPPLY
1  60              30                  
2 -50             -55                  
3  NA              NA                  
4   1              60                  
5   1              60                  
6   1              NA

    STUDYID        DAYS_MED_INDEX                        DRUG_NAME     
 Min.   :    103   Min.   :-4834   METFORMIN HCL              : 99710  
 1st Qu.: 261123   1st Qu.:  329   METFORMIN HCL 500 MG TABLET: 35672  
 Median : 513041   Median :  971   METFORMIN HCL ER           : 34704  
 Mean   : 549969   Mean   : 1302   ACTOS                      : 26168  
 3rd Qu.: 830052   3rd Qu.: 1970   GLIMEPIRIDE                : 24595  
 Max.   :1256011   Max.   : 7264   METFORMIN    TAB 500MG     : 23854  
                                   (Other)                    :465041  
    NDC_CODE            STRENGTH      DISPENSE_AMOUNT   NUMBER_OF_DAYS_SUPPLY
 Min.   :2.751e+06   Min.   :   0.0   Min.   :-473.00   Min.   :-100.00      
 1st Qu.:9.310e+07   1st Qu.:  10.0   1st Qu.:  30.00   1st Qu.:  30.00      
 Median :5.911e+08   Median : 100.0   Median :  60.00   Median :  30.00      
 Mean   :2.621e+10   Mean   : 307.8   Mean   :  69.85   Mean   :  44.06      
 3rd Qu.:6.276e+10   3rd Qu.: 500.

Now we know that this dataset has 7 variables:

* Studyid(`STUDYID`)
* Days from or to the index date (`DAYS_MED_INDEX`)
* Name of the drug (`DRUG_NAME`)
* National drug code (`NDC_CODE`)
* Strength of the medication (`STRENGTH`)
* Number of pills given out (`DISPENSE_AMOUNT`)
* Number of days supply (`NUMBER OF DAYS SUPPLY`)

The medication data is useful for examining the drugs that are being prescribed to patients and we first examine the frequency distribution of the drugs using the `table` function. <br>
`as.data.frame` formats the table, `heads` function limits the rows to first 6 rows. 

In [2]:
head(as.data.frame(table(medication$DRUG_NAME)))

Var1                           Freq
1 53746017810                      2 
2 ACARBOSE                       279 
3 ACARBOSE     TAB 100MG          18 
4 ACARBOSE     TAB 100MGACARBOSE   3 
5 ACARBOSE     TAB 25MG           57 
6 ACARBOSE     TAB 25MGACARBOSE   27

What we find here is that given that our drugs are identified by the specific generic drug and formulation combination, we see multiple entries for the same type of drug e.g. Acarbose, Acarbose Tab 100mg, Acarbose Tab 25 mg etc. Further, even variations in spellings result in the drugs being considered distinct. This highly granular information is of little use from a research standpoint when examining most research questions. It would be helpful to aggregate these drugs into more meaningful categories such as combining acarbose, actoplus and all other similar drugs into the "Anti-diabetic medication" category. In order to do so we will merge this data with a cross walk that provides categories of drugs.

The crosswalk dataset to be used is the `<filename>.dta` available in the curricular materials. 

The crosswalk was created in a different statistical software called Stata and exists as a `.dta` file. Thus, a different library will be used to bring the cross walk into R.

In the following cells, we will load the `haven` library, use it to read the `<filename>.dta` into R, and view the first few rows.

In [3]:
#Install the haven package using following code if not already installed.
#install.packages("haven") 
library("haven")

#Loading .dta (Stata file)
unindc <- read_dta("unindc.dta") #reading the STATA file 
cat("Dimensions (rows columns/variables):",dim(unindc)) #Queries number of rows and columns/variables
head(unindc) 

Dimensions (rows columns/variables): 429473 9

ndc         drug                                                   rxncui
1 00002000001 Ergocalciferol 1.25 Mg Oral Capsule [Deltalin]          4018 
2 00002000002 Ergocalciferol 1.25 Mg Oral Capsule [Deltalin]          4018 
3 00002001402 Flurandrenolide 0.004 Mg/Sqcm Medicated Tape [Cordran]  4500 
4 00002001801 Capreomycin 500 Mg/Ml Injectable Solution [Capastat]   78903 
5 00002001901 Cycloserine 250 Mg Oral Capsule [Seromycin]             3007 
6 00002003401 Cephalexin 500 Mg Oral Tablet [Keftab]                  2231 
  concept_name    cscode  drug_class     
1 Vitamin D2      A11CC01 Ergocalciferol 
2 Vitamin D2      A11CC01 Ergocalciferol 
3 Flurandrenolone D07AC07 Fludroxycortide
4 Capreomycin     J04AB30 Capreomycin    
5 Cycloserine     J04AB01 Cycloserine    
6 Cephalexin      J01DB01 Cefalexin      
  sgthr                                       
1 Vitamins                                    
2 Vitamins                                    
3 Corticosteroids, Dermatological Preparations
4 Antimycobacterials                          
5 Antimycobacterials                          
6 Antibacterials For Systemic Use             
  sgphr                                     
1 Vit A And D, Incl. Combinations Of The Two
2 Vit A And D, Incl. Combinations Of The Two
3 Corticosteroids, Plain                    
4 Drugs For Treatment Of Tuberculosis       
5 Drugs For Treatment Of Tuberculosis       
6 Other Beta-Lactam Antibacterials          
  sgchem                             
1 Vitamin D And Analogues            
2 Vitamin D And Analogues            
3 Corticosteroids, Potent (Group Iii)
4 Antibiotics                        
5 Antibiotics                        
6 First-Generation Cephalosporins

The crosswalk file: `<filename.dta>` consists of 429,473 rows and the following 9 variables:

* National Drug code (`ndc`)
* Drug name (`drug`)
* RxNorm clinical drug concept (`rxncui`)
* Concept name (`concept_name`)
* WHO/ATC code (`cscode`)
* Drug class (`drug_class`)
* Therapeutic subgroup (`sgthr`)
* Pharmacological subgroup (`sgphr`)
* Chemical subgroup (`sgchem`) 

For our purposes, we can use the various subgroups (`sgthr`-`sgchem`) to aggregate medications. Note that the `drug_class` variable while valid for aggregation does not provide meaningful categories.

The `ndc` variable uniquely identifies all observations in the crosswalk and can thus be used for a many-to-one merge.

### Merge
We now merge this file with our medications file using the `ndc` variable which is common to both the croswalk and the T2D medications file. However, this variable is named differently in both datasets, while in the crosswalk it is called `ndc`, in the medications dataset it is called `NDC_CODE`. Merging requires that the variable to be merged on have the same name and this is the first step in merging.

In [4]:
#Matching NDC column name for merging 
print("List of variables in medication data in the order they appear")
names(medication) #query names of the variables, output is the names of the variables in order they appear in data
colnames(unindc)[1] <- "NDC_CODE" #renames first var in crosswalk to match the corresponding variable name in medications data
print("ndc variable in crosswalk changed to match variable name in medications dataset")
names(unindc)

[1] "List of variables in medication data in the order they appear"


[1] "STUDYID"               "DAYS_MED_INDEX"        "DRUG_NAME"            
[4] "NDC_CODE"              "STRENGTH"              "DISPENSE_AMOUNT"      
[7] "NUMBER_OF_DAYS_SUPPLY"

[1] "ndc variable in crosswalk changed to match variable name in medications dataset"


[1] "NDC_CODE"     "drug"         "rxncui"       "concept_name" "cscode"      
[6] "drug_class"   "sgthr"        "sgphr"        "sgchem"

We `merge` by matching all drug codes in the medications file with corresponding drugs in the crosswalk as follows:

In [5]:
medication_unindc <- merge(x=medication, y=unindc, by= "NDC_CODE", all.x= TRUE)
head(medication_unindc)
cat("Dimensions (rows columns/variables):",dim(medication_unindc))
summary(medication_unindc)

NDC_CODE STUDYID DAYS_MED_INDEX DRUG_NAME                 STRENGTH
1 2751001  131681  2059           HUMALOG                    NA     
2 2751001  131681  2920           HUMALOG                    NA     
3 2751001  395666  1020           HUMALOG                    NA     
4 2751001     686  4647           HUMALOG 100 UNITS/ML VIAL 100     
5 2751001  513260  2728           HUMALOG 100 UNITS/ML VIAL 100     
6 2751001    1038   757           HUMALOG      INJ 100/ML    NA     
  DISPENSE_AMOUNT NUMBER_OF_DAYS_SUPPLY drug rxncui concept_name cscode
1 20              30                    NA   NA     NA           NA    
2 20              30                    NA   NA     NA           NA    
3 NA              30                    NA   NA     NA           NA    
4 10              30                    NA   NA     NA           NA    
5 90              90                    NA   NA     NA           NA    
6  1              60                    NA   NA     NA           NA    
  drug_class sgthr sgphr sgchem
1 NA         NA    NA    NA    
2 NA         NA    NA    NA    
3 NA         NA    NA    NA    
4 NA         NA    NA    NA    
5 NA         NA    NA    NA    
6 NA         NA    NA    NA

Dimensions (rows columns/variables): 709744 15

    NDC_CODE            STUDYID        DAYS_MED_INDEX 
 Min.   :2.751e+06   Min.   :    103   Min.   :-4834  
 1st Qu.:9.310e+07   1st Qu.: 261123   1st Qu.:  329  
 Median :5.911e+08   Median : 513041   Median :  971  
 Mean   :2.621e+10   Mean   : 549969   Mean   : 1302  
 3rd Qu.:6.276e+10   3rd Qu.: 830052   3rd Qu.: 1970  
 Max.   :6.865e+10   Max.   :1256011   Max.   : 7264  
                                                      
                       DRUG_NAME         STRENGTH      DISPENSE_AMOUNT  
 METFORMIN HCL              : 99710   Min.   :   0.0   Min.   :-473.00  
 METFORMIN HCL 500 MG TABLET: 35672   1st Qu.:  10.0   1st Qu.:  30.00  
 METFORMIN HCL ER           : 34704   Median : 100.0   Median :  60.00  
 ACTOS                      : 26168   Mean   : 307.8   Mean   :  69.85  
 GLIMEPIRIDE                : 24595   3rd Qu.: 500.0   3rd Qu.:  90.00  
 METFORMIN    TAB 500MG     : 23854   Max.   :2700.0   Max.   :6000.00  
 (Other)                    :465041   NA's   :495

### Match statistics

In [6]:
cat("Drugs that could not be matched:", sum(is.na(medication_unindc$drug)))  #this shows non-matching count  
cat("\nTotal number of unique drugs matched:",length(unique(medication_unindc$drug))) # number of unique drugs matched 
as.data.frame(table(medication_unindc$sgthr)) #The list of the drugs

Drugs that could not be matched: 405949
Total number of unique drugs matched: 87

Var1                                      Freq  
1 Beta Blocking Agents                          64
2 Blood Substitutes And Perfusion Solutions      8
3 Drugs Used In Diabetes                    303456
4 Other Gynecologicals                         162
5 Pancreatic Hormones                          105

As mentioned above, only 87 unique drugs were matched resulting in a total of 405,949 drugs being <b>not</b> matched to a corresponding drug in the crosswalk. Why does this happen?

The ndc is a 11-digit code, i.e. it consists solely of numbers. If we examine the medications file we find that the ndc code is a string variable with leading zeroes in some codes. When we read the medications file, R tries to automatically assign variable types based on the contents of each column. Given that ndc codes are all numbers, R designates the ndc code in the medication dataset as a numeric variable dropping the leading zeros. 

Why does the crosswalk file not experience the same error?
The crosswalk file was a Stata file and includes class settings in the meta data. When the haven library imports this file, this allows R to preserve the class settings for all variables. 

In [7]:
cat("NDC class in crosswalk dataset:",class(unindc$NDC_CODE))
cat("\nNDC class in medications dataset:", class(medication$NDC_CODE))

NDC class in crosswalk dataset: character
NDC class in medications dataset: numeric

To fix the error above, we force R to read the medication file while preserving the drug codes as a string or character variable.

In [8]:
medication <- read.csv("medication.csv", header= TRUE, 
                      colClasses=c("character", "numeric", "character", "character", rep("numeric",3)) )
summary(medication)
head(medication, n= 10)

   STUDYID          DAYS_MED_INDEX   DRUG_NAME           NDC_CODE        
 Length:709744      Min.   :-4834   Length:709744      Length:709744     
 Class :character   1st Qu.:  329   Class :character   Class :character  
 Mode  :character   Median :  971   Mode  :character   Mode  :character  
                    Mean   : 1302                                        
                    3rd Qu.: 1970                                        
                    Max.   : 7264                                        
                                                                         
    STRENGTH      DISPENSE_AMOUNT   NUMBER_OF_DAYS_SUPPLY
 Min.   :   0.0   Min.   :-473.00   Min.   :-100.00      
 1st Qu.:  10.0   1st Qu.:  30.00   1st Qu.:  30.00      
 Median : 100.0   Median :  60.00   Median :  30.00      
 Mean   : 307.8   Mean   :  69.85   Mean   :  44.06      
 3rd Qu.: 500.0   3rd Qu.:  90.00   3rd Qu.:  60.00      
 Max.   :2700.0   Max.   :6000.00   Max.   : 660.00      
 N

STUDYID DAYS_MED_INDEX DRUG_NAME              NDC_CODE    STRENGTH
1  103     2757           GLIPIZIDE 5 MG TABLET  51079081020 NA      
2  239     6871           NOVOLIN N              00169183411 NA      
3  239      811           HUMULIN N              00002831501 NA      
4  239     3751           HUMULIN N    INJ U-100 00002831501 NA      
5  239     3819           HUMULIN R    INJ U-100 00002821501 NA      
6  239     1128           HUMULIN N              00002831501 NA      
7  239     1219           HUMULIN N              00002831501 NA      
8  239     2076           HUMULIN N    INJ U-100 00002831501 NA      
9  239     3819           HUMULIN N    INJ U-100 00002831501 NA      
10 239      811           HUMULIN N              00002831501 NA      
   DISPENSE_AMOUNT NUMBER_OF_DAYS_SUPPLY
1   60              30                  
2  -50             -55                  
3   NA              NA                  
4    1              60                  
5    1              60                  
6    1              NA                  
7    1              NA                  
8    1              NA                  
9    1              60                  
10  -1              NA

Now that we fixed the NDC code and confirm that leading zeros are preserved we proceed with the merge as before.

In [9]:
 
medication_unindc <- merge(x= medication, y= unindc, by= "NDC_CODE", all.x= TRUE)
head(medication_unindc)
summary(medication_unindc)


NDC_CODE    STUDYID DAYS_MED_INDEX DRUG_NAME                      STRENGTH
1 00002751001 700552    80           HUMALOG 100 UNITS/ML VIAL       NA     
2 00002751001 695895   -54           HUMALOG 100 UNITS/ML VIAL       NA     
3 00002751001 248910  3447           HUMALOG                        100     
4 00002751001 465949  3490           HUMALOG      INJ 100/MLHUMALOG  NA     
5 00002751001 460314  2048           HUMALOG 100 UNITS/ML VIAL       NA     
6 00002751001 141427   257           HUMALOG      INJ 100/ML         NA     
  DISPENSE_AMOUNT NUMBER_OF_DAYS_SUPPLY
1 -10             -10                  
2  10              10                  
3   3              90                  
4  10              30                  
5  30              30                  
6   1              30                  
  drug                                                    rxncui
1 Insulin Lispro 100 Unt/Ml Injectable Solution [Humalog] 86009 
2 Insulin Lispro 100 Unt/Ml Injectable Solution [Humalog] 86009 
3 Insulin Lispro 100 Unt/Ml Injectable Solution [Humalog] 86009 
4 Insulin Lispro 100 Unt/Ml Injectable Solution [Humalog] 86009 
5 Insulin Lispro 100 Unt/Ml Injectable Solution [Humalog] 86009 
6 Insulin Lispro 100 Unt/Ml Injectable Solution [Humalog] 86009 
  concept_name           cscode  drug_class     sgthr                 
1 Insulin, Lispro, Human A10AB04 Insulin Lispro Drugs Used In Diabetes
2 Insulin, Lispro, Human A10AB04 Insulin Lispro Drugs Used In Diabetes
3 Insulin, Lispro, Human A10AB04 Insulin Lispro Drugs Used In Diabetes
4 Insulin, Lispro, Human A10AB04 Insulin Lispro Drugs Used In Diabetes
5 Insulin, Lispro, Human A10AB04 Insulin Lispro Drugs Used In Diabetes
6 Insulin, Lispro, Human A10AB04 Insulin Lispro Drugs Used In Diabetes
  sgphr                  sgchem                                           
1 Insulins And Analogues Insulins And Analogues For Injection, Fast-Acting
2 Insulins And Analogues Insulins And Analogues For Injection, Fast-Acting
3 Insulins And Analogues Insulins And Analogues For Injection, Fast-Acting
4 Insulins And Analogues Insulins And Analogues For Injection, Fast-Acting
5 Insulins And Analogues Insulins And Analogues For Injection, Fast-Acting
6 Insulins And Analogues Insulins And Analogues For Injection, Fast-Acting

   NDC_CODE           STUDYID          DAYS_MED_INDEX   DRUG_NAME        
 Length:709744      Length:709744      Min.   :-4834   Length:709744     
 Class :character   Class :character   1st Qu.:  329   Class :character  
 Mode  :character   Mode  :character   Median :  971   Mode  :character  
                                       Mean   : 1302                     
                                       3rd Qu.: 1970                     
                                       Max.   : 7264                     
                                                                         
    STRENGTH      DISPENSE_AMOUNT   NUMBER_OF_DAYS_SUPPLY     drug          
 Min.   :   0.0   Min.   :-473.00   Min.   :-100.00       Length:709744     
 1st Qu.:  10.0   1st Qu.:  30.00   1st Qu.:  30.00       Class :character  
 Median : 100.0   Median :  60.00   Median :  30.00       Mode  :character  
 Mean   : 307.8   Mean   :  69.85   Mean   :  44.06                         
 3rd Qu.: 500.0   3rd Q

In [10]:
# Drugs field that did not match (NA count) is 12 now (successfull matches are 720,889)
cat("Drugs that could not be matched:", sum(is.na(medication_unindc$drug)))  #this shows non-matching count  
cat("\nTotal number of unique drugs matched:",length(unique(medication_unindc$drug))) # number of unique drugs matched 

Drugs that could not be matched: 12
Total number of unique drugs matched: 215

After fixing the error, only 12 drugs are not matched compared to over 400,000 before. Also, compared to 87 unique drugs that were matched now we have 215 unique drugs that were matched.

### Let's take a look at the diabetes drugs prescribed

In [11]:
as.data.frame(table(medication_unindc$sgthr))
as.data.frame(table(medication_unindc$sgchem[medication_unindc$sgthr == "Drugs Used In Diabetes"]))
as.data.frame(table(medication_unindc$drug_class[medication_unindc$sgthr == "Drugs Used In Diabetes"]))


Var1                                      Freq  
1 Antihypertensives                             10
2 Beta Blocking Agents                          64
3 Blood Substitutes And Perfusion Solutions    151
4 Drugs Used In Diabetes                    708757
5 Lipid Modifying Agents                        12
6 Other Gynecologicals                         162
7 Pancreatic Hormones                          576

Var1                                               Freq  
1  Alpha Glucosidase Inhibitors                         1172
2  Biguanides                                         292325
3  Dipeptidyl Peptidase 4 (Dpp-4) Inhibitors           48889
4  Insulins & Analogues For Inject, Intermediate-Act   11132
5  Insulins And Analogues For Inhalation                   8
6  Insulins And Analogues For Injection, Fast-Acting   34710
7  Insulins And Analogues For Injection, Long-Acting   54117
8  Other Blood Glucose Lowering Drugs, Excl. Insulins  21974
9  Sulfonamides, Urea Derivatives                     172943
10 Thiazolidinediones                                  71487

Var1              Freq  
1  Acarbose             902
2  Alogliptin           138
3  Canagliflozin       1835
4  Chlorpropamide       152
5  Combinations          18
6  Exenatide           8485
7  Glibenclamide      38758
8  Glimepiride        58665
9  Glipizide          75346
10 Insulin (Human)    15741
11 Insulin Aspart     14752
12 Insulin Detemir     9938
13 Insulin Glargine   43390
14 Insulin Glulisine   1320
15 Insulin Lispro     14808
16 Linagliptin         2763
17 Liraglutide         6993
18 Metformin         292325
19 Miglitol             270
20 Nateglinide         2629
21 Pioglitazone       56595
22 Pramlintide          448
23 Repaglinide         1584
24 Rosiglitazone      14888
25 Saxagliptin         5728
26 Sitagliptin        40260
27 Tolbutamide           22
28 Troglitazone           4

Finally, the frequency table provides much more useful information when we use the therapeutic subgroups. How does this compare to the pharmacological subgroups and the chemical subgroup?

### Let's save the enriched medication file (in R format)
To load the file, use the following code: <br>
`load(file = "medication_unindc.RDATA")`

In [12]:
save(medication_unindc , file ="medication_unindc.RDATA")
file.info("medication_unindc.RDATA")

size    isdir mode mtime              
medication_unindc .RDATA 5358112 FALSE 660  2018-10-25 09:50:40
                         ctime               atime               uid    gid 
medication_unindc .RDATA 2018-10-25 09:50:40 2018-10-25 09:47:56 498945 1297
                         uname    grname
medication_unindc .RDATA ukirbiyi T15

== End of Ex 1 ==